### PR TITLE
Implement slice.

### PR DESCRIPTION
**Pull request by BruceDai**
According to the [slice op explanation](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-slice)
```
starts: a sequence of long. The starting indices to slice of the corresponding axes of the input shape. 
A negative index value is interpreted as counting back from the end. For example, the value -1

sizes: a sequence of long. The lengths to slice of the corresponding axes of the input shape. 
The length value of -1 selects all the remaining elements from the starting index of the given axis.
```

I added two slice tests to test `negative starts` and `sizes having -1` scenarios.

@huningxin PTAL, thanks.

### CODE REVIEW

File: src/slice.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {MLSliceOptions} options
+ * @return {Tensor}
+ */
+export function slice(input, starts, sizes, options = {}) {
+  const rank = input.rank;
+  const allStarts = new Array(rank).fill(0);
+  let axes = options.axes;
+  if (axes) {
+    if (axes.length > rank) {
+      throw new Error('The length of axes is invalid.');
```
Comment (huningxin) [inline src/slice.js:33]: Could the error message be more specific? like `The length of axes ${axes.length} is larger than rank ${rank}`.

File: src/slice.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    } else {
+      for (let i = 0; i < axes.length; ++i) {
```
Comment (huningxin) [inline src/slice.js:35]: `for (const axis of axes)`

File: src/slice.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        if (!Number.isInteger(axes[i]) || axes[i] >= rank || axes[i] < -rank) {
+          throw new Error('The value of axes is invalid.');
```
Comment (huningxin) [inline src/slice.js:37]: more specific error message

File: src/slice.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        }
+      }
+    }
+  } else {
+    axes = [...Array(rank).keys()];
+  }
+  const axesLen = axes.length;
+  if (starts.length !== axesLen) {
+    throw new Error('The length of starts is invalid.');
```
Comment (huningxin) [inline src/slice.js:46]: ditto

File: src/slice.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  }
+  if (sizes.length !== axesLen) {
+    throw new Error('The length of sizes is invalid.');
```
Comment (huningxin) [inline src/slice.js:49]: ditto

File: src/slice.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  }
+  const outputShape = input.shape.slice();
+  for (let j = 0; j < axesLen; ++j) {
```
Comment (huningxin) [inline src/slice.js:52]: For top level loop, could the iterator be i? `for (let i = 0; i < axesLen; ++i)`

File: src/slice.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ * Produce a slice of the input tensor.
+ * @param {Tensor} input
+ * @param {Array} starts
+ * @param {Array} sizes
+ * @param {MLSliceOptions} options
+ * @return {Tensor}
+ */
+export function slice(input, starts, sizes, options = {}) {
+  const rank = input.rank;
+  const allStarts = new Array(rank).fill(0);
```
Comment (huningxin) [inline src/slice.js:29]: `startsForAllAxes`

File: src/slice.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  const axesLen = axes.length;
+  if (starts.length !== axesLen) {
+    throw new Error('The length of starts is invalid.');
+  }
+  if (sizes.length !== axesLen) {
+    throw new Error('The length of sizes is invalid.');
+  }
+  const outputShape = input.shape.slice();
+  for (let j = 0; j < axesLen; ++j) {
+    const slicedDim = axes[j] >= 0 ? axes[j] : axes[j] + rank;
```
Comment (huningxin) [inline src/slice.js:53]: axis

File: src/slice.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    const slicedDimSize = input.shape[slicedDim];
```
Comment (huningxin) [inline src/slice.js:54]: size

File: src/slice.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    const slicedStart = starts[j];
```
Comment (huningxin) [inline src/slice.js:55]: start

File: src/slice.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    allStarts[slicedDim] = slicedStart >= 0 ? slicedStart : slicedStart + slicedDimSize;
+    if (!Number.isInteger(slicedStart) || slicedStart >= slicedDimSize ||
+        slicedStart < -slicedDimSize) {
+      throw new Error('The value of starts is invalid.');
```
Comment (huningxin) [inline src/slice.js:59]: error message ditto

File: src/slice.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    } else {
+      const slicedSize = sizes[j];
+      if (Number.isInteger(slicedSize) && slicedSize >= 0) {
+        if (slicedStart >= 0) {
+          if (slicedStart + slicedSize > slicedDimSize) {
+            throw new Error('The value of sizes is invalid.');
```
Comment (huningxin) [inline src/slice.js:65]: ditto

File: src/slice.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+          } else {
+            outputShape[slicedDim] = slicedSize;
+          }
+        } else {
+          if (slicedStart + slicedSize > 0) {
+            throw new Error('The value of sizes is invalid.');
```
Comment (huningxin) [inline src/slice.js:71]: ditto

File: src/slice.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+          } else {
+            outputShape[slicedDim] = slicedSize;
+          }
+        }
+      } else {
+        if (slicedSize !== -1) {
+          throw new Error('The value of sizes is invalid.');
```
Comment (huningxin) [inline src/slice.js:78]: ditto

File: src/slice.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        } else {
+          outputShape[slicedDim] = slicedStart >= 0 ? slicedDimSize - slicedStart : -slicedStart;
+        }
+      }
+    }
+  }
+  const output = new Tensor(outputShape);
+  for (let k = 0; k < sizeOfShape(outputShape); ++k) {
```
Comment (huningxin) [inline src/slice.js:86]: k -> outputIndex

File: src/slice.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    const loc = output.locationFromIndex(k);
+    const selectedInputLoc = updateLocation(loc, allStarts);
```
Comment (huningxin) [inline src/slice.js:88]: `updateLocation` is only used once, please outline it here.

**Review by huningxin — COMMENTED** [review]
good work, thanks @BruceDai , I left some comments, please take a look.

**Comment by BruceDai** [comment]
Thanks @huningxin, I‘ve updated the code, please take a look again, thanks.

File: src/slice.js [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    const start = starts[i];
+    if (!Number.isInteger(start)) {
+      throw new Error(`Invalid starts value ${start}, it should be an integer.`);
+    }
+    startsForAllAxes[axis] = start >= 0 ? start : start + size;
+    if (start >= size || start < -size) {
+      throw new Error(`Invalid starts value ${start}, it shoule be in the interval ` +
+        `[${-size}, ${size}).`);
+    } else {
+      const slicedSize = sizes[i];
```
Comment (huningxin) [inline src/slice.js:66]: size

File: src/slice.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+      if (!Number.isInteger(slicedSize)) {
+        throw new Error(`Invalid sizes value ${slicedSize}, it should be an integer.`);
+      }
+      if (Number.isInteger(slicedSize) && slicedSize >= 0) {
```
Comment (huningxin) [inline src/slice.js:70]: Why do you need to check integer again?

File: src/slice.js [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+        } else {
+          outputShape[axis] = start >= 0 ? size - start : -start;
+        }
+      }
+    }
+  }
+  const output = new Tensor(outputShape);
+  for (let outputIndex = 0; outputIndex < sizeOfShape(outputShape); ++outputIndex) {
+    const loc = output.locationFromIndex(outputIndex);
+    const selectedInputLoc = updateLocation(loc, startsForAllAxes);
```
Comment (huningxin) [inline src/slice.js:99]: I mean remove `updateLocation` and implement its algorithm directly here.

File: src/slice.js (hunk 17 above)
Comment (BruceDai) [reply src/slice.js:66]: The variable `size` was used above for 
```js
const size = input.shape[axis];
```
so here needs another variable, thanks.

File: src/slice.js (hunk 18 above)
Comment (BruceDai) [reply src/slice.js:70]: Thanks. I missed to update it here.

File: src/slice.js (hunk 17 above)
Comment (BruceDai) [reply src/slice.js:66]: @huningxin How about rename `size` to 'length'

```js
const length = input.shape[axis];
```

, then use `size` here?

```js
const size = sizes[i];
```

Comment (huningxin) [reply src/slice.js:66]: Good point. I missed that. Then `sliceSize`, it means the size of the slice. Thanks.

**Comment by BruceDai** [comment]
Thanks @huningxin, I updated the commit, please take a look again.

**Review by huningxin — APPROVED** [review]

**Pull request merged**